[DTensor][fix] fix is_tensor_shardable to correctly handle Replicate placement

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118134
* #118132
* __->__ #117726

**Summary**
Previously DTensor sharding plans filter (i.e. `is_tensor_shardable()`) cannot correctly handle the case where the input `DTensor` has 0 dimension. This filter should return `True` if the sharding placement on 0 dimension is `Replicate` even if `tensor dim < num of shards` on that dimension in which case `tensor dim == 0` and `num of shards == 1`.

In this PR we also noticed a behavior discrepancy of `torch.addmm`. See #118131 

**Test Plan**
```
pytest test/distributed/_tensor/test_dtensor_ops.py -s -k addmm
pytest test/distributed/_tensor/test_dtensor_ops.py -s -k mm_cpu_float32
CUDA_VISIBLE_DEVICES="" pytest test/distributed/_tensor/test_matrix_ops.py -s -k empty_operand
pytest test/distributed/_tensor/test_matrix_ops.py -s -k empty_operand
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

XilunWu committed:
[DTensor][issue] addmm cannot handle 0-element input tensor

[ghstack-poisoned]

wconstab commented on test/distributed/_tensor/test_matrix_ops.py:
```diff
@@ -39,6 +39,25 @@ def test_addmm(self):
+        local_res = torch.addmm(input_tensor, tensor_to_shard, tensor_to_replicate)
+        print(f"rank={self.rank}, dist_res._spec={dist_res._spec}, dist_res._local_tensor={dist_res._local_tensor}")
+        print(f"rank={self.rank}, dist_res={dist_res.full_tensor()}, local_res={local_res}")
+        self.assertNotEqual(dist_res.full_tensor(), local_res)
```
I'm not following this logic, after your fix shouldn't the distributed tensor match the local tensor?

XilunWu replied:
It should be, so I mark this PR as WIP :-(

Unfortunately I haven't found a good solution on this issue. Initially I thought Wanchao's PR #115525 fixed this case but it turned out to be a NO...

I will leave that to another PR since it's caused by another issue in DTensor.

XilunWu committed:
Update on "[DTensor][issue] addmm cannot handle 0-element input tensor"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu commented:
Interesting, the CI shows a different signal from a local run:
command: `pytest test/distributed/_tensor/test_matrix_ops.py -s -k test_addmm_empty_operand`

wanchaol reviewed (approved):
Nice fix! changes LGTM.

Heads up we can have a follow up PR to re-enable op db tests to test this behavior as we disabled this recently in https://github.com/pytorch/pytorch/pull/118045.

XilunWu committed:
Update on "[DTensor][fix] fix is_tensor_shardable to correctly handle Replicate placement"


**Summary**
Previously DTensor sharding plans filter (i.e. `is_tensor_shardable()`) cannot correctly handle the case where the input `DTensor` has 0 dimension. This filter should return `True` if the sharding placement on 0 dimension is `Replicate` even if `tensor dim < num of shards` on that dimension in which case `tensor dim == 0` and `num of shards == 1`.

**Test Plan**
pytest test/distributed/_tensor/test_dtensor_ops.py -s -k addmm
pytest test/distributed/_tensor/test_dtensor_ops.py -s -k mm_cpu_float32

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][fix] fix is_tensor_shardable to correctly handle Replicate placement"


**Summary**
Previously DTensor sharding plans filter (i.e. `is_tensor_shardable()`) cannot correctly handle the case where the input `DTensor` has 0 dimension. This filter should return `True` if the sharding placement on 0 dimension is `Replicate` even if `tensor dim < num of shards` on that dimension in which case `tensor dim == 0` and `num of shards == 1`.

In this PR we also noticed a behavior discrepancy of `torch.addmm`. See #118131 

**Test Plan**
pytest test/distributed/_tensor/test_dtensor_ops.py -s -k addmm
pytest test/distributed/_tensor/test_dtensor_ops.py -s -k mm_cpu_float32
CUDA_VISIBLE_DEVICES="" pytest test/distributed/_tensor/test_matrix_ops.py -s -k empty_operand
pytest test/distributed/_tensor/test_matrix_ops.py -s -k empty_operand

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu commented:
failed inductor test is unrelated
